lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
-                this.useSpecificConfig = loadConfig(useConfig);
-            } else {
-                this.useSpecificConfig = loadConfig(path.resolve(this.options.cwd, useConfig));
+            if (!(isResolvable(useConfig) || isResolvable(`eslint-config-${useConfig}`) || useConfig.charAt(0) === "@")) {
+                useConfig = path.resolve(this.options.cwd, useConfig);
             }
+            this.useSpecificConfig = ConfigFile.loadCached(useConfig);
+        }
+
+        if (this.options.plugins) {
+            Plugins.loadAll(this.options.plugins);
         }
+
+        // Empty values in configs don't merge properly
+        const cliConfigOptions = {
+            env: this.env,
+            rules: this.options.rules,
+            globals: this.globals,
+            parserOptions: this.parserOptions,
+            plugins: this.options.plugins
+        };
+
+        this.cliConfig = {};
+        Object.keys(cliConfigOptions).forEach(configKey => {
+            const value = cliConfigOptions[configKey];
+
+            if (value) {
+                this.cliConfig[configKey] = value;
+            }
+        });
     }
 
     /**
-     * Build a config object merging the base config (conf/eslint-recommended),
-     * the environments config (conf/environments.js) and eventually the user
-     * config.
-     * @param {string} filePath a file in whose directory we start looking for a local config
-     * @returns {Object} config object
+     * Gets the personal config object from user's home directory.
+     * @returns {Object} the personal config object (null if there is no personal config)
+     * @private
      */
-    getConfig(filePath) {
-        const directory = filePath ? path.dirname(filePath) : this.options.cwd;
-        let config,
-            userConfig;
-
-        debug(`Constructing config for ${filePath ? filePath : "text"}`);
+    getPersonalConfig() {
+        if (typeof this.personalConfig === "undefined") {
+            let config;
 
-        config = this.cache[directory];
+            if (PERSONAL_CONFIG_DIR) {
+                const filename = ConfigFile.getFilenameForDirectory(PERSONAL_CONFIG_DIR);
 
-        if (config) {
-            debug("Using config from cache");
-            return config;
+                if (filename) {
+                    debug("Using personal config");
+                    config = ConfigFile.loadCached(filename);
+                }
+            }
+            this.personalConfig = config || null;
         }
 
-        // Step 1: Determine user-specified config from .eslintrc.* and package.json files
+        return this.personalConfig;
+    }
+
+    /**
+     * Builds a hierarchy of config objects, including the base config, all local configs from the directory tree,
+     * and a config file specified on the command line, if applicable.
+     * @param {string} directory a file in whose directory we start looking for a local config
+     * @returns {Object[]} The config objects
```

Comment:
Can you update this comment to specify how the config objects are ordered in the array? (Are the config objects with higher precedence placed at the beginning of the array, or the end?)